src/shogun/lib/any.h

```diff
@@ -1222,9 +1223,12 @@ namespace shogun {
 
 	template <typename T>
 	inline void register_casts() {
-		if constexpr (std::is_base_of<CSGObject, typename std::remove_pointer<T>::type>::value)
+		using Derived = std::remove_pointer_t<T>;
+		if constexpr (std::is_base_of_v<CSGObject, Derived>)
 		{
 			Any::register_caster<T, CSGObject*>([] (T value) { return dynamic_cast<CSGObject*>(value); });
+			if constexpr (!std::is_same_v<std::nullptr_t, base_type<Derived>>)
+				Any::register_caster<T, base_type<Derived>*>([] (T value) { return dynamic_cast<base_type<Derived>*>(value); });
```

Comment:
doh... this fails with msvc 19.16, but not with 19.20 or after:
https://godbolt.org/z/8DjP0R

Reply:
if i change the definition of base_type to
```
using base_type = typename std::decay_t<decltype(find_base<Derived>(type_list<Types...>{}))>;
```

and then do `dynamic_cast<base_type<Derived>::type *>(...);` it works.... anybody any ideas? :D

Reply:
```using base_type = typename decltype(find_base<Derived>(sg_interface{}))::type;```
solved the problem... lemme know if you feel that that could cause some troubles :S